object-store/src/main/java/org/hypertrace/config/objectstore/ConfigObjectScopeResolver.java

```diff
@@ -0,0 +1,131 @@
+package org.hypertrace.config.objectstore;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Resolver class to resolve configs based on scope
+ *
+ * @param <T> Raw config type
+ * @param <U> Resolved config type
+ * @param <S> Scope
+ */
+public abstract class ConfigObjectScopeResolver<T extends Message, U extends Message, S> {
+
+  protected final Optional<U> defaultConfig;
+
+  public ConfigObjectScopeResolver(U defaultConfig) {
+    this.defaultConfig = Optional.of(defaultConfig);
+  }
+
+  public ConfigObjectScopeResolver() {
+    this.defaultConfig = Optional.empty();
+  }
+
+  public List<U> getAllResolvedConfigData(List<T> configDataList) {
+    Map<S, T> configDataMap = getConfigDataMap(configDataList);
+    return configDataList.stream()
+        .flatMap(configData -> resolveConfig(extractScope(configData), configDataMap).stream())
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  public Optional<U> getResolvedData(List<T> configDataList, S scope) {
+    Map<S, T> configDataMap = getConfigDataMap(configDataList);
+    return resolveConfig(scope, configDataMap);
+  }
+
+  /**
+   * Method to extract scope object from given config object
+   *
+   * @param configData
+   * @return Scope object
+   */
+  protected abstract S extractScope(T configData);
+
+  /**
+   * Method to return scopes with which the given scope needs to be resolved in increasing order of
+   * priority For example, API scope would return a list {customer, env, service, api}
```

Comment:
nit
```suggestion
   * priority For example, API scope could return a list {tenant, service, api}
```

Keeping this relevant to HT, also theres no requirement that a service has to live above an api, for example, so just loosening the wording slightly.